app/docs/0.11.x/plugin-development/entities-cache.md

```diff
@@ -15,195 +15,275 @@ chapter: 7
-* The datastore would also be affected by an increase of load, potentially crashing or slowing down the datastore, which in turn would affect every Kong node.
-
-To avoid querying the datastore every time, we can cache custom entities in-memory on the node, so that frequent entity lookups don't trigger a datastore query every time (only the first time), but happen in-memory, which is much faster and reliable that querying it from the datastore (especially under heavy load).
-
-<div class="alert alert-warning">
-  <strong>Note:</strong> When caching custom entities in-memory, then you also need to provide an invalidation mechanism, implemented in the "hooks.lua" file.
-</div>
+Your plugin may need to frequently access custom entities (explained in the
+[previous chapter]({{page.book.previous}})) on every request and/or response.
+Usually, loading them once and caching them in-memory dramatically improves
+the performance while making sure the datastore is not stressed with an
+increased load.
+
+Think of an api-key authentication plugin that needs to validate the api-key on
+every request, thus loading the custom credential object from the datastore on
+every request. When the client provides an api-key along with the request,
+normally you would query the datastore to check if that key exists, and then
+either block the request or retrieve the Consumer ID to identify the user. This
+would happen on every request, and it would be very inefficient:
+
+* Querying the datastore adds latency on every request, making the request
+  processing slower.
+* The datastore would also be affected by an increase of load, potentially
+  crashing or slowing down, which in turn would affect every Kong
+  node.
+
+To avoid querying the datastore every time, we can cache custom entities
+in-memory on the node, so that frequent entity lookups don't trigger a
+datastore query every time (only the first time), but happen in-memory, which
+is much faster and reliable that querying it from the datastore (especially
+under heavy load).
 
 ---
 
 ### Caching custom entities
 
-Once you have defined your custom entities, you can cache them in-memory in your code by requiring the `database_cache` dependency:
+Once you have defined your custom entities, you can cache them in-memory in
+your code by using the `singletons.cache` module provided by Kong:
 
 ```
-local cache = require "kong.tools.database_cache"
+local singletons = require "kong.singletons"
+local cache = singletons.cache
 ```
 
 There are 2 levels of cache:
 
-1. Lua memory cache (local to an nginx worker)
+1. L1: Lua memory cache (local to an nginx worker)
    This can hold any type of Lua value.
-2. Shared memory cache - SHM (local to an nginx node, but shared between all workers)
-   This can only hold scalar values, and hence requires (de)serialization.
+2. L2: Shared memory cache - SHM (local to an nginx node, but shared between all
+   workers). This can only hold scalar values, and hence requires
```

Comment:
style suggestion:

```
2. L2: Shared memory cache (SHM) - local to an nginx node, but shared between all
   workers. This can only hold scalar values, and hence requires
```